### PR TITLE
Paper: Circumventing The Linker: Using SciPy's BLAS and LAPACK Within Cython (Henriksen)

### PR DESCRIPTION
**Pull request by insertinterestingnamehere**
Here's a draft of my proceedings paper. Thanks for the help putting this together!

[Edit by jaberg] Paperbot: http://nipy.bic.berkeley.edu:5000/download/4

### CODE REVIEW

**Comment by katyhuff** [comment]
Hi There, 
I wanted to let you know that your paper is not currently building in [our build server](http://nipy.bic.berkeley.edu:5000/#). The error message is thus:

```
• Updating proceedings build tools in /home/katyhuff/procbuild/procbuild/../cache/scipy_proceedings...
Already up-to-date.
• Check out paper repository...
Cloning into '/tmp/tmpaWXxvf'...
Installing proceedings build tools...
• Build the paper...
Building: ian_henriksen
Traceback (most recent call last):
File "publisher/build_paper.py", line 179, in 
build_paper(paper_id)
File "publisher/build_paper.py", line 164, in build_paper
rst2tex(in_path, out_path)
File "publisher/build_paper.py", line 84, in rst2tex
settings_overrides=settings)
File "/usr/lib/python2.7/site-packages/docutils/core.py", line 414, in publish_string
enable_exit_status=enable_exit_status)
File "/usr/lib/python2.7/site-packages/docutils/core.py", line 662, in publish_programmatically
output = pub.publish(enable_exit_status=enable_exit_status)
File "/usr/lib/python2.7/site-packages/docutils/core.py", line 219, in publish
output = self.writer.write(self.document, self.destination)
File "/usr/lib/python2.7/site-packages/docutils/writers/__init__.py", line 80, in write
self.translate()
File "/usr/lib/python2.7/site-packages/docutils/writers/latex2e/__init__.py", line 246, in translate
self.document.walkabout(visitor)
File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
if child.walkabout(visitor):
File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
if child.walkabout(visitor):
File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
if child.walkabout(visitor):
File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 166, in walkabout
visitor.dispatch_visit(self)
File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 1882, in dispatch_visit
return method(node)
File "/tmp/tmpaWXxvf/publisher/writer/__init__.py", line 374, in visit_literal_block
lexer = get_lexer_by_name(node.attributes['language'])
File "/usr/lib/python2.7/site-packages/pygments/lexers/__init__.py", line 80, in get_lexer_by_name
raise ClassNotFound('no lexer for alias %r found' % _alias)
pygments.util.ClassNotFound: no lexer for alias u'Fortran' found
Error building paper /tmp/tmpaWXxvf/papers/ian_henriksen. Aborting.
Error code 1
```


**You can try fixing stuff and then pressing the refresh button to the right of your paper title on the build server page until you get it building.**

**Comment by insertinterestingnamehere** [comment]
@katyhuff Thanks for pointing it out!

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+   Up until now the standard practice in many packages for using BLAS and LAPACK has been to link each Python extension directly against the libraries needed.
+   Each module that calls these low-level libraries directly has had to link against them independently.
+
+   Cython has existing machinery that allows C-level declarations to be shared between Cython-compiled extension modules without linking against the original libraries.
+   If used carefully, this functionality can be used to export Cython-level APIs exactly mirroring existing C and Fortran interfaces that can be used without any additional linking.
+   The Cython BLAS and LAPACK API in SciPy makes it so that the same BLAS and LAPACK libraries that were used to compile SciPy can be used in Python extension modules via Cython.
+
+   In this talk I will demonstrate how to create and use these APIs for both Fortran and C libraries in a platform-independent manner.
+   I will also discuss how using these techniques mitigate the effects of dependency creep both within and between Python packages.
+
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: IMO this abstract is probably like... twice as long as should be, about?

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+Cython has also provided a similar sort of functionality by allowing C-level APIs to be exported between Cython modules.
+
+
+The Cython API for BLAS and LAPACK
+----------------------------------
+
+Over the last year, a significant amount of work has been devoted to exposing the BLAS and LAPACK libraries within SciPy at the Cython level.
+The primary goals of providing such an interface are twofold:
+* Making the low-level routines in BLAS and LAPACK more readily available to users.
+* Reducing the dependency burden on third party packages.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: This isn't formatting as a list in the pdf output.

I'd probably skip the bulleted list and just write it inline: "twofold: First, making the low-level routines, ..., and second, reducing the dependency burden..."

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+Using the new Cython API, users can now dynamically load the BLAS and LAPACK libraries used to compile SciPy without having to actually link against the original BLAS and LAPACK libraries or include the corresponding headers.
+Modules that use the new API also no longer need to worry about which BLAS or LAPACK library is used.
+If the correct versions of BLAS and LAPACK were used to compile SciPy, the correct versions will be used by the extension module.
+
+BLAS and LAPACK proved to be particularly good candidates for a Cython API, resulting in several additional benefits:
+* Python modules that use the Cython BLAS/LAPACK API no longer need to link statically to provide binary installers.
+* The custom ABI wrappers and patches used in SciPy to provide a more stable and uniform interface across different BLAS/LAPACK libraries and  Fortran compilers are no longer needed for third party extensions.
+* The naming schemes used within BLAS and LAPACK make it easy to write type-dispatching versions of BLAS and LAPACK routines using Cython's fused types.
+
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: This bullet list also is failing to bullet.

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+       # Axes lengths are all equal, so n can be reused.
+       # Note that all arguments are passed as pointers.
+       dgemm(&trans, &trans, &n, &n, &n, &al, &a[0,0],
+             &n, &b[0,0], &n, &be, &c[0,0], &n)
+       # Print the results to show they match np.dot.
+       print np.array(c)
+       print np.dot(a, b)
+
+If these wrappers are needed in an extension module written in C, C++, or another low-level language, a small Cython shim can be used to export the needed functions.
+Since Cython uses Python's capsule objects internally for the cimport mechanism, it is also possible to extract function pointers directly from the module's `__pyx_capi__` dictionary and cast them to the needed type without writing the extra shim.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: You want double-backticks around `__pyx_capi__`, not single backticks.

I have no idea how to type an example of what I mean into this github form, so hopefully that is clear.

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+The process of exposing a Cython binding for a function or variable in an existing library is relatively simple.
+First, as an example, consider the simple C file
+
+.. code-block:: c
+
+   // myfunc.c
+   double f(double x, double y){
+   return x * x - x * y + 3 * y;
+   }
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: Please use standard spacing/indentation!
```
double f(double x, double y) {
    return x * x - x * y + 3 * y;
}
```

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+with the corresponding header file
+
+.. code-block:: c
+
+   // myfunc.h
+   double f(double x, double y);
+
+This library can be compiled by running `clang -c myfunc.c -o myfunc.o`.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: Double-backticks again. I'll assume that you'll do a search through the document for other instances and stop worrying about pointing them out :-)

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+   # cy_myfunc.pyx
+   # Use a file-level directive to link
+   # against the compiled object.
+   # distutils: extra_link_args = ['myfunc.o']
+   cdef extern from 'myfunc.h':
+       double f(double x, double y) nogil
+   # Declare both the external function and
+   # the Cython function as nogil so they can be
+   # without any Python operations
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: "so that they can be without" seems to be missing a word.

**Comment by njsmith** [comment]
Overall comments:

I really like the paper, but I thought the structure was a bit wobbly. In particular the introduction is really meandering and kind of talks around the main point a lot without ever quite saying it.

I'd restructure the paper to start with a very concrete example, something like:

* Here's a piece of code that doesn't run fast enough. The usual advice is to switch to compiled code.

* Okay, here's the same code rewritten in Cython. But to make it fast we need a direct call to BLAS. Which means, first we have to use the right include file, and then figure out which BLAS we're using and which name mangling is in use, and switch to buil̸ding ẃit́h ͟n͜ųmp̷y͏.di͠stut͡ils̛,͝ and́ at͘ r̨uņti͠m͘e w̵e̴ h͝a͟v̴e ̧t͠o ́f̸i̴nd ̶tḩe͡ ̕lib͜rár͝y͏, a̘̺̤̰̕n̬̦̯̮̹̘d̡͕̼ͅ i̶̜̭̠f̘̞̙ ͉̲̗ẁ̬͎e̤̜̙̝ ͖͡w̱̠̫̦̤̝a̺̥͢ṋ͇̯͓ͅt̨͉ ̬t̛̰̟̰ͅo̢ ̭̠̼Ḏ͕I̝͇̹̼̰ͅͅS͔̯T̛R̰͉̹͍̮I̶̺͙͕B̛̗̻̣̖̳U̬̫̩̞͇͚͟T̗͘E̗̟̲͉ ̭̣̲͙͔̭ò̲͕͍̱̪͇̺u͇͕͖͙͞r ̡͚̣͖͓c̴͕̪͍̻od̬͈͕͕̫͚̮̀e͝ ͙̹̰̪̺̰̲ț̪͉̮̯̪̠h̼̱͕͍̰e̲̘n̛̞̫̻̻̲̙͇ ̱h̭̘̐̑̔ͨa̷̶̹̣͙̺̜̺̟̱͐ͤ͡ͅh̴̼̲̹̿̀a̛̙̦̮̤̠͉͉͐̄̌͋́͟h̲͉̹̯̝̙͎̎ͧ́͡a̪̘̦͓̩̜ͣ̉ͧ̊ͯͤͩͅ ͙̖̻͉͔̖͕ͥ͌ͤͪ̈́͐̔͢ͅh̀ͦ̅̓̋͆̏҉̩͎̙͕̤̝͓͡ȅ̠͎̲̲̯̳̫̝͊̎͒̿̃ͮ̚ ̫̯͎̝͈͔̬͖̩̈̍͛̓͂̋͒c̘̣̮̞̃̍̃ͭͩ̀o̷̙̤̩̫̹̍̊ͮ̉͠ṃ̬̻̑̀̾ͬ̒͂̇́ë̗̹̰̤̲͓͚̯ͣs̤͋̀͂͒ͪ̂ͦ̚͢ H̷̡̲̗͙̫̝̰̩̆̀̀̈́ͫ̇ͩ̀̽ͣ̿͌̚Ȩ̨̯̪̫̲̙̣̹̳͉̼͌́ͭ̑͋ͣ̓̎̃ͫ͛̒̄͜ ̧͆̂̑́ͮ̉ͬͫ͂̈̈̍́̃̒ͧ͋͘҉͖͙̲̹͙͙̦̀ͅC̡̙̘̤̠͎͕̩̖͓̝̺̻̗̻̣̟̪̯̩ͫͯ̃̒̐̊ͭ̃̃́͞Oͫ̄͋̈̐̃͂̅ͮ͑̔̇̚͏̧͕͉͕̫̼̯̣͓̟̫̹̳̤̙̬͜M̷̗͓̰̝͇̺̳̏̀̔̈̋͒͋̑̆́́ͅE̵̷͓̫͍̲̦̞ͯͧͭ́ͩ̐ͫ̐̊ͥ̚͠Sͬ̓͑ͭͪ̅̌̍̃̿ͫͧͤ̇̃ͫ̌́̓͡҉̶̰̖̪̘̯̟̥̦ͅ

* Oh but wait, Cython has this really clever tool where Cython functions exported from one module can be accessed directly from another Cython module while totally skipping the system linker and using normal Python lookup rules instead.

* (Optional: give a sidebar explaining how this works internally, for those who are curious and to raise the general level of knowledge of such things.)

* Check it out, scipy now uses this to re-export the BLAS and LAPACK functions, so we can skip all the unspeakable horrors above and just write the following: `...`. Wasn't that nicer?

* Here's some examples of how you can use the same trick to export routines from your own libraries that someone else might need to call from a Cython extension without C overhead.

I think this would be pretty much the same content, but much more accessible.

**Comment by prabhuramachandran** [comment]
This is interesting work that promises to be very useful to those using Python for HPC.

- I agree with @njsmith on his comments completely.

- As @njsmith points out, paper could be relatively easily rewritten for more
  clarity. The motivation while clear could be made better with a simple
  example that shows the problem first, and how the current `linalg.cython_blas`
  module nicely solves the problem with an elaboration of how it works.

**Comment by insertinterestingnamehere** [comment]
@njsmith @prabhuramachandran, thank you both very much for your comments. I'll get a revised version up soon. The changed structure sounds like a great idea.

**Comment by jaberg** [comment]
Great exchange so far! Looks like we're good to go when @insertinterestingnamehere uploads revisions.

**Comment by insertinterestingnamehere** [comment]
Okay, I've uploaded a restructured version. Thanks for the comments thus far!

**Comment by prabhuramachandran** [comment]
@insertinterestingnamehere -- looks much better now, thanks!  @jaberg: This addresses my major concerns and the paper reads much nicer now.  If @njsmith agrees, I think this is RTM.

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+---------------------------------------------------------------------
+Circumventing The Linker: Using SciPy's BLAS and LAPACK Within Cython
+---------------------------------------------------------------------
+
+.. class:: abstract
+
+   BLAS, LAPACK, and other libraries like them have formed the underpinnings of much of the scientific stack in Python.
+   Until now, the standard practice in many packages for using BLAS and LAPACK has been to link each Python extension directly against the libraries needed.
+   Each module that calls these low-level libraries directly has had to link against them independently.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: I'd add just a tiny bit more here, like "...and this creates substantial obstacles to building and distributing such modules."

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+             &n, &info)
+
+Though this process for calling an external function from a library isn't particularly difficult, the setup file for the Python module now must find a proper LAPACK installation.
+If there are several different versions of LAPACK present, a suitable one must be chosen.
+The proper headers and libraries must be found, and, if at all possible, binary incompatibilities between compilers must be avoided.
+If the desired routine isn't a part of one of the existing C interfaces, then it must be called via the Fortran ABI and the name mangling schemes used by different Fortran compilers must be taken into account.
+All of the code needed to do this must also be maintained so that it continues to work with new versions of the different operating systems, compilers, and BLAS and LAPACK libraries.
+
+An effective solution to this unusually painful problem is to have existing Python modules provide access to the low-level libraries that they use.
+NumPy has provided some of this sort of functionality for BLAS and LAPACK by making it so that the locations of the system's BLAS and LAPACK libraries can be found using NumPy's distutils module.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: I'd be clear here that `numpy.distutils` can be helpful when building a module, but it's only for build time, not run time; and in particular, it doesn't necessarily help once you've copied the compiled module onto another system that might have a different BLAS and LAPACK in a different place.

File: papers/ian_henriksen/cython_blas_lapack_api.rst [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+Cython has also provided a similar sort of functionality by allowing C-level APIs to be exported between Cython modules.
+In fact, the existing machinery in Cython can be used to expose functions and variables from existing libraries.
```
Comment (njsmith) [inline papers/ian_henriksen/cython_blas_lapack_api.rst]: I'd say "Normally, this machinery is only used to let Cython functions in one module call other Cython functions in another module. But if we're clever, it turns out we can also use this mechanism as a general way to expose functions and variables from existing libraries, while circumventing all the problems that arise when trying to use the standard linker." This move -- of taking a cute little Cython-internal hack and reconsidering it as a general replacement for the OS linker -- is really the key contribution of the paper IMO, so you should make it super clear :-).

**Comment by njsmith** [comment]
I agree, I like this new version a lot :-). Some minor comments but otherwise looks good to me.

**Comment by insertinterestingnamehere** [comment]
Alright, I believe I've addressed all the current comments. Is there anything else?

**Comment by njsmith** [comment]
I'm happy.

**Comment by insertinterestingnamehere** [comment]
Great! Thanks for the helpful reviews!

**Pull request merged**